userspace/libscap/scap_procs.c

```diff
@@ -700,6 +700,83 @@ int32_t scap_proc_fill_exe_writable(scap_t* handle, struct scap_threadinfo* tinf
 	return SCAP_SUCCESS;
 }
 
+void scap_proc_fill_exe_upper_layer(scap_t* handle, struct scap_threadinfo* tinfo, const char *procdirname)
+{
+	char buf[SCAP_MAX_PATH_SIZE];
+	char upperdir[SCAP_MAX_PATH_SIZE];
+	char exe[SCAP_MAX_PATH_SIZE];
+	char line[SCAP_MAX_ARGS_SIZE];
+	FILE *mounts_file;
+	struct stat upperdir_exe_stat;
+	int ret;
+
+	//
+	// Open mounts file
+	// 
+	snprintf(buf, sizeof(buf), "%smounts", procdirname);
+	mounts_file = fopen(buf, "r");
+	if(!mounts_file)
+	{
+		return;
+	}
+
+	//
+	// Check if there is an overlayfs mounted in root. 
+	// This is used to understand if the process is within a container.
+	// If this is the case, the first line of the mounts file contains the string "overlay / overlay" 
+	// and we can retrieve the upperdir path for this container.
+	//
+	fgets(line, sizeof(line), mounts_file);
+	if(strstr(line, "overlay / overlay"))
+	{
+		char *start, *end;
+
+		start = strstr(line, "upperdir=") + sizeof("upperdir=") - 1;
+		if(!start)
+		{
+			return;
+		}
+
+		end = strstr(start, ",");
+		if(!end)
+		{
+			return;
+		}
+		*end = '\0';
+
+		snprintf(upperdir, sizeof(upperdir), "%s", start);
+	}
+	else
+	{
+		fclose(mounts_file);
+		return;
+	}
+
+	fclose(mounts_file);
+
+	//
+	// Read link of the executable file
+	//
+	snprintf(buf, sizeof(buf), "%s%s", procdirname, "exe");
+	ret = readlink(buf, exe, sizeof(exe));
+	if(ret < 0)
+	{
+		return;
+	}
+	*(exe + ret) = '\0';
+
+	//
+	// Check if the upperdir directory contains the executable file 
+	// and set upper_layer accordingly
+	//
+	strncat(upperdir, exe, sizeof(upperdir) - strlen(upperdir) - 1);
+	ret = stat(upperdir, &upperdir_exe_stat);
```

Comment:
If I understand the description right, that would be the part that fails in case Falco runs in a container. I believe it would need the right host prefix to work in that case if the appropriate directory is mounted. The other workaround would be to use `/proc/1/root/...` but I agree it's not the cleanest way of doing that.

Reply:
I found out the function `scap_get_host_root` and I think it solves the problem. Falco uses the environment variable `HOST_ROOT`, which is set to `/host` in the official docker image. I believe that if we mount directories such as `/var/lib/docker`, their path will follow this pattern too, ending up in `/host/var/lib/docker`. So just prepending it to each path should work, thank you! :)

Reply:
I am going to try it!

Reply:
I have tested it, and it now works also when Falco runs in a container if `/var/lib/docker` is mounted. Thank you!